ui/job-view/details/tabs/TabsPanel.jsx

```diff
@@ -0,0 +1,241 @@
+      const { tabIndex, showAutoclassifyTab } = this.state;
+      const { perfJobDetail } = this.props;
+      const nextIndex = tabIndex + 1;
+      const tabCount = TabsPanel.getTabNames(!!perfJobDetail.length, showAutoclassifyTab).length;
+      this.setState({ tabIndex: nextIndex < tabCount ? nextIndex : 0 });
+    });
+
+    this.setTabIndex = this.setTabIndex.bind(this);
+  }
+
+  componentWillUnmount() {
+    this.selectNextTabUnlisten();
+  }
+
+  setTabIndex(tabIndex) {
+    this.setState({ tabIndex });
+  }
+
+  render() {
+    const {
+      jobDetails, jobLogUrls, logParseStatus, suggestions, errors, pinJob, user, bugs,
+      bugSuggestionsLoading, selectedJob, perfJobDetail, repoName, jobRevision,
+      classifications, togglePinBoardVisibility, isPinBoardVisible, pinnedJobs, addBug,
+      classificationTypes, lvFullUrl, reftestUrl, $injector,
+    } = this.props;
+    const { showAutoclassifyTab, tabIndex } = this.state;
+    const countPinnedJobs = Object.keys(pinnedJobs).length;
+
+    return (
+      <div id="tabs-panel" className="job-tabs-divider">
+        <Tabs
+          selectedTabClassName="selected-tab"
+          selectedIndex={tabIndex}
+          onSelect={this.setTabIndex}
+        >
+          <TabList className="tab-headers">
+            <span className="tab-header-tabs">
+              <Tab>Job Details</Tab>
+              <Tab>Failure Summary</Tab>
+              {showAutoclassifyTab && <Tab>Failure Classification</Tab>}
+              <Tab>Annotations</Tab>
+              <Tab>Similar Jobs</Tab>
+              {!!perfJobDetail.length && <Tab>Performance</Tab>}
+            </span>
+            <span id="tab-header-buttons" className="details-panel-controls pull-right">
+              <span
+                id="pinboard-btn"
+                className="btn pinboard-btn-text"
+                onClick={togglePinBoardVisibility}
+                title={isPinBoardVisible ? 'Close the pinboard' : 'Open the pinboard'}
+              >PinBoard
+                {!!countPinnedJobs && <div
+                  title={`You have ${countPinnedJobs} job${countPinnedJobs > 1 ? 's' : ''} pinned`}
+                  className={`pin-count-group ${countPinnedJobs > 99 ? 'pin-count-group-3-digit' : ''}`}
+                >
+                  <div
+                    className={`pin-count-text ${countPinnedJobs > 99 ? 'pin-count-group-3-digit' : ''}`}
+                  >{countPinnedJobs}</div>
+                </div>}
+                <span
+                  className={`fa ${isPinBoardVisible ? 'fa-angle-down' : 'fa-angle-up'}`}
+                />
+              </span>
+              <span
+                onClick={() => this.$rootScope.$emit(thEvents.clearSelectedJob)}
+                className="btn details-panel-close-btn"
+              ><span className="fa fa-times" /></span>
+            </span>
+          </TabList>
+          <TabPanel>
```

Comment:
Interesting idea, I'll give it a shot and see how that works.  :)

Reply:
In trying to do this, it ends up that ``react-tabs`` checks that there are a matching number of ``TabPanel`` to ``Tab`` components.  If I move ``TabPanel`` to inside the tab, like ``FailureSummaryTab`` then it doesn't detect there are a matching number of ``Tab`` to ``TabPanel``.  If I have ``FailureSummaryTab`` extend ``TabPanel`` rather than ``React.Component`` then it DOES work.  However, only for some of the tabs.  It appears to break tab navigation.

I haven't looked too deeply, but I fear that extending ``TabPanel`` may cause state in our component to potentially interfere with ``TabPanel``.  It may also be that it's causing some interference with applying classes to show/hide in some rendering code in ``TabPanel``.

I'm thinking this may not be an avenue worth exploring further at this time.  What do you think?

Reply:
Yeah that sounds fine :-)

Btw I can't see the stylesheet being imported anywhere? (https://github.com/reactjs/react-tabs#styling) Would using that reduce the boilerplate we need in our own stylesheet? (If not, don't worry)